File: src/main/java/com/rnave/studily/flashcard/Sm2.java
package com.rnave.studily.flashcard;

/**
 * SM-2 spaced-repetition scheduling (SuperMemo 2, as popularized by Anki),
 * with Anki's four answer buttons mapped onto SM-2 quality scores.
 */
public final class Sm2 {

    public static final double MIN_EASE_FACTOR = 1.3;

    public enum Grade {
        AGAIN(0),
        HARD(3),
        GOOD(4),
        EASY(5);

        private final int quality;

        Grade(int quality) {
            this.quality = quality;
        }
    }

    public record Result(int repetitions, double easeFactor, int intervalDays) {}

    private Sm2() {}

    public static Result review(int repetitions, double easeFactor, int intervalDays, Grade grade) {
        if (grade == Grade.AGAIN) {
            // Failed recall: restart the repetition sequence; SM-2 leaves the ease factor
            // unchanged on failure. Interval 0 keeps the card due now, so it comes back
            // within the same study session.
            return new Result(0, easeFactor, 0);
        }

        int q = grade.quality;
        double ef = Math.max(MIN_EASE_FACTOR,
                easeFactor + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02)));

        int reps = repetitions + 1;
        int interval;
        if (reps == 1) {
            interval = 1;
        } else if (reps == 2) {
            interval = 6;
        } else {
            interval = (int) Math.round(intervalDays * ef);
        }
        return new Result(reps, ef, interval);
    }
}

File: frontend/src/features/learn/sm2.ts
import type { Flashcard, ReviewGrade } from "../../types";

const QUALITY: Record<ReviewGrade, number> = { AGAIN: 0, HARD: 3, GOOD: 4, EASY: 5 };

export function isDue(card: Flashcard, now = Date.now()): boolean {
  return !card.dueAt || new Date(card.dueAt).getTime() <= now;
}

export function nextIntervalDays(card: Flashcard, grade: ReviewGrade): number {
  if (grade === "AGAIN") return 0;
  const reps = (card.repetitions ?? 0) + 1;
  if (reps === 1) return 1;
  if (reps === 2) return 6;
  const q = QUALITY[grade];
  const ef = Math.max(1.3, (card.easeFactor ?? 2.5) + (0.1 - (5 - q) * (0.08 + (5 - q) * 0.02)));
  return Math.round((card.intervalDays ?? 0) * ef);
}

export function intervalLabel(days: number): string {
  if (days <= 0) return "now";
  if (days < 30) return `${days}d`;
  if (days < 365) return `${Math.round(days / 30)}mo`;
  return `${(days / 365).toFixed(1)}y`;
}
